File: Vesti/Vesti/HomeViewController.m
//
//  HomeViewController.m
//  Vesti
//
//  Created by Djuro Alfirevic on 2/3/17.
//  Copyright © 2017 Djuro Alfirevic. All rights reserved.
//

#import "HomeViewController.h"
#import "Article.h"
#import "ArticleTableViewCell.h"

//#define URL @"http://www.brzevesti.net/api/news"
static NSString *const URL = @"http://www.brzevesti.net/api/news";

@interface HomeViewController() <UITableViewDataSource, UITableViewDelegate>
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (strong, nonatomic) NSMutableArray *itemsArray;
@end

@implementation HomeViewController

#pragma mark - Private API

- (void)loadArticles {
    // Grand Central Dispatch (GCD)
    dispatch_queue_t downloadQueue = dispatch_queue_create("ArticlesDownloader", NULL);
    dispatch_async(downloadQueue, ^{
        // This code will be executed on BACKGROUND thread
        NSURL *url = [NSURL URLWithString:URL];
        NSData *data = [NSData dataWithContentsOfURL:url];
        NSDictionary *dictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];

        self.itemsArray = [[NSMutableArray alloc] init];

        for (NSDictionary *articleDictionary in dictionary[@"news"]) {
            Article *article = [[Article alloc] initWithDictionary:articleDictionary];
            [self.itemsArray addObject:article];
        }

        dispatch_async(dispatch_get_main_queue(), ^{
            // This code will be executed on MAIN thread
            [self.tableView reloadData];
        });
    });
}

#pragma mark - View lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];

    [self loadArticles];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.itemsArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    ArticleTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell"];
    
    cell.article = self.itemsArray[indexPath.row];

    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

@end

File: VestiS/VestiS/HomeViewController.swift
//
//  HomeViewController.swift
//  VestiS
//
//  Created by Djuro Alfirevic on 3/29/17.
//  Copyright © 2017 Djuro Alfirevic. All rights reserved.
//

import UIKit
import SwiftyJSON

class HomeViewController: UIViewController {

    // MARK: - Properties
    @IBOutlet weak var tableView: UITableView!
    var articles = [Article]()

    // MARK: - Public API
    func loadArticlesWithSwiftyJSON() {
        guard isConnected() else {
            return
        }

        if let url = URL(string: apiUrl) {
            let configuration = URLSessionConfiguration.default
            let request = URLRequest(url: url)
            let session = URLSession(configuration: configuration)
            let task = session.dataTask(with: request, completionHandler: { [unowned self] (data, response, error) in

                if let error = error {
                    print("Error: \(error.localizedDescription)")
                }

                if let data = data {
                    let json = JSON(data)

                    DispatchQueue.main.async { [unowned self] in
                        if let news = json["news"].array {
                            for newsJson in news {
                                let article = Article(with: newsJson)
                                self.articles.append(article)
                            }

                            self.tableView.reloadData()
                        }
                    }
                }
            })
            
            task.resume()
        }
    }

    func loadArticles() {
        guard isConnected() else {
            return
        }

        if let url = URL(string: apiUrl) {
            let configuration = URLSessionConfiguration.default
            let request = URLRequest(url: url)
            let session = URLSession(configuration: configuration)
            let task = session.dataTask(with: request, completionHandler: { [unowned self] (data, response, error) in

                if let error = error {
                    print("Error: \(error.localizedDescription)")
                }

                if let data = data {
                    // try, try?, try!

                    do {
                        let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
                        self.parseArticles(json: json)
                    } catch {
                        print("Error: \(error.localizedDescription)")
                    }
                }
            })

            task.resume()
        }
    }

    func parseArticles(json: [String: Any]?) {
        guard let json = json else {
            return
        }

        DispatchQueue.main.async { [unowned self] in
            if let news = json["news"] as? [[String: Any]] {
                for newsJson in news {
                    let article = Article(with: newsJson)
                    self.articles.append(article)
                }

                self.tableView.reloadData()
            }
        }
    }

    // MARK: - View lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.tableFooterView = UIView()
        //loadArticles()
        loadArticlesWithSwiftyJSON()
    }

    // MARK: - Segue
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let indexPath = tableView.indexPathForSelectedRow else {
            return
        }

        let article = articles[indexPath.row]

        // 1. version.
        if segue.destination is WebViewController {
            let toViewController = segue.destination as! WebViewController
            toViewController.urlString = article.url
        }

        // 2. version.
        if let toViewController = segue.destination as? WebViewController {
            toViewController.urlString = article.url
        }
    }

}

extension HomeViewController: UITableViewDataSource, UITableViewDelegate {

    // MARK: - UITableViewDataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return articles.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell") as! ArticleTableViewCell
        cell.article = articles[indexPath.row]

        return cell
    }

    // MARK: - UITableViewDelegate
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }

}
